plugins/chain_plugin/get_info_db.cpp

```diff
@@ -0,0 +1,159 @@
+      std::string           server_version_string;
+      std::string           server_full_version_string;
+
+      void store_info_common(const std::shared_ptr<get_info_db::get_info_results>& info) {
+         assert(info);
+
+         // fixed part
+         info->server_version             = server_version;
+         info->chain_id                   = chain_id;
+         info->server_version_string      = server_version_string;
+         info->server_full_version_string = server_full_version_string;
+
+         // chain head part
+         const auto& head = controller.head();
+         if (head.is_valid()) {
+            info->head_block_id       = head.id();
+            info->head_block_num      = block_header::num_from_id(info->head_block_id);
+            info->head_block_time     = head.block_time();
+            info->head_block_producer = head.producer();
+         }
+
+         // fork_db part
+         if (controller.fork_db_has_root()) {
+            info->fork_db_head_block_id        = controller.fork_db_head().id();
+            info->fork_db_head_block_num       = block_header::num_from_id(*info->fork_db_head_block_id);
+            info->earliest_available_block_num = controller.earliest_available_block_num();
+         }
+
+         // resource_limits
+         const auto& rm = controller.get_resource_limits_manager();
+         info->virtual_block_cpu_limit = rm.get_virtual_block_cpu_limit();
+         info->virtual_block_net_limit = rm.get_virtual_block_net_limit();
+         info->block_cpu_limit         = rm.get_block_cpu_limit();
+         info->block_net_limit         = rm.get_block_net_limit();
+         info->total_cpu_weight        = rm.get_total_cpu_weight();
+         info->total_net_weight        = rm.get_total_net_weight();
+      }
+
+      void store_info() {
+         std::shared_ptr<get_info_db::get_info_results> info = std::make_shared<get_info_db::get_info_results>();
+
+         store_info_common(info);
+
+         if (controller.fork_db_has_root()) {
+            info->last_irreversible_block_id   = controller.fork_db_root().id();
+            info->last_irreversible_block_num  = block_header::num_from_id(info->last_irreversible_block_id);
+            info->last_irreversible_block_time = controller.fork_db_root().block_time();
+         }
+
+         std::atomic_store(&info_cache, info);  // replace current cache safely
+      }
+
+      void store_info(const chain::signed_block_ptr& block, const block_id_type& id) {
+         std::shared_ptr<get_info_db::get_info_results> info = std::make_shared<get_info_db::get_info_results>();
+
+         store_info_common(info);
+
+         info->last_irreversible_block_id   = id;
+         info->last_irreversible_block_num  = block_header::num_from_id(info->last_irreversible_block_id);
+         info->last_irreversible_block_time = block->timestamp;
+
+         std::atomic_store(&info_cache, info);  // replace current cache safely
```

Comment:
This is not lock free. Most likely worse than using `std::atomic<std::shared_ptr>`.

Reply:
Should be similar, I think. It is deprecated, but since we don't have `std::atomic<std::shared_ptr>` probably the best. Certainly seems better to me than the boost atomic shared ptr with its spin lock.
https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

Reply:
I agree. It is simpler and is locked shorter than mutex.

Will add an issue to remind us to change to atomic shared_ptr once it is supported by compilers.